core/src/main/java/contacts/core/Where.kt

```diff
@@ -403,108 +468,256 @@ private class RawContactsTableWhere(whereString: String) : Where<RawContactsFiel
-    )
+    copyWithNewFieldType { field ->
+        when (field) {
+            Fields.RawContact.Id -> RawContactsFields.Id
+            else -> field // no substitution
+        }
+    }
 
 // endregion
 
 // region Where
 
 /**
- * Each where expression is paired with its mimetype because the contacts Data table uses
- * generic column names (e.g. data1, data2, etc) using the column 'mimetype' to distinguish
- * the type of data in that generic column.
+ * The WHERE clause of a database query made up of a specific type of [Field] ([T]).
  *
- * For example, querying for contacts with name LIKE 'john' AND address LIKE 'colorado';
+ * ## Developer notes
  *
- * ```
- * WHERE (data1 = 'john' AND mimetype = 'vnd.android.cursor.item/name')
- *   AND (data1 = 'colorado' AND mimetype = 'vnd.android.cursor.item/postal-address_v2')
- * ```
+ * The type [T] is not exactly "used" in this class itself. Rather, it is used for adding type
+ * restrictions when constructing instances at compile time. For example, this allows us to create a
+ * function that takes in a Where of GroupsField. The caller of that function can then only provide
+ * a Where composed of one or more GroupsField.
  *
- * This is important because if the mimetypes are not paired with the query;
+ * The type [T] is not enforced within the class itself in order to support mutating functions
+ * such as [inRawContactsTable] and [inContactsTable]. This will allow us to construct a Where<X>
+ * from a Where<Y>.
  *
- * ```
- * WHERE (data1 = 'johnson' AND data1 = 'colorado')
- * ```
+ * ### Binary tree structure
+ *
+ * The form "lhs operator rhs" naturally forms a binary tree. A where can only be constructed in
+ * two different ways (hence the private constructor and two secondary constructors).
  *
- * The above will never match any row because 'johnson' = 'colorado' is never true.
+ * - Base case: (lhs=FieldHolder, rhs=ValueHolder)
+ *     - THis can have a parent or siblings but cannot have children. AKA a leaf node.
+ *     - This can be the only node (the root not) in a tree.
+ * - Recursive case: (lhs=WhereHolder, rhs=WhereHolder)
+ *     - This can have a parent or siblings and MUST have children.
+ *     - This cannot be the only node in a tree though it can be the root node.
+ *
+ * For example,
+ *
+ *                                 WhereHolder
+ *          WhereHolder                                    WhereHolder
+ *   FieldHolder  ValueHolder               WhereHolder                   WhereHolder
+ *                                   FieldHolder  ValueHolder       FieldHolder  ValueHolder
+ *
+ *  With this in mind, we can do some cool stuff like in [copyWithNewFieldType]!
  */
-private fun where(field: Field, operator: String, value: Any?, options: String? = null): String {
-    var where = "${field.columnName} $operator ${value.toSqlString()}"
+class Where<out T : Field> private constructor(
```

Comment:
The primary constructor is private. Two secondary constructors are exposed to ensure that the structure of the binary tree is upheld.